op-challenger/game/keccak/verifier.go

```diff
@@ -3,33 +3,63 @@ package keccak
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 
 	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/fetcher"
 	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/matrix"
 	keccakTypes "github.com/ethereum-optimism/optimism/op-challenger/game/keccak/types"
+	"github.com/ethereum-optimism/optimism/op-service/sources/batching"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
+	lru "github.com/hashicorp/golang-lru/v2"
 )
 
+const validPreimageCacheSize = 500
+
+type VerifierPreimageOracle interface {
+	fetcher.Oracle
+	GetProposalTreeRoot(ctx context.Context, block batching.Block, ident keccakTypes.LargePreimageIdent) (common.Hash, error)
+}
+
 type Fetcher interface {
 	FetchInputs(ctx context.Context, blockHash common.Hash, oracle fetcher.Oracle, ident keccakTypes.LargePreimageIdent) ([]keccakTypes.InputData, error)
 }
 
 type PreimageVerifier struct {
 	log     log.Logger
 	fetcher Fetcher
+
+	// knownValid caches the merkle tree roots that have been confirmed as valid.
+	// Invalid roots are not cached as those preimages will be ignored once the challenge is processed.
+	knownValid *lru.Cache[common.Hash, bool]
 }
 
 func NewPreimageVerifier(logger log.Logger, fetcher Fetcher) *PreimageVerifier {
+	// Can't error because size is hard coded
+	cache, _ := lru.New[common.Hash, bool](validPreimageCacheSize)
 	return &PreimageVerifier{
-		log:     logger,
-		fetcher: fetcher,
+		log:        logger,
+		fetcher:    fetcher,
+		knownValid: cache,
 	}
 }
 
-func (v *PreimageVerifier) CreateChallenge(ctx context.Context, blockHash common.Hash, oracle fetcher.Oracle, preimage keccakTypes.LargePreimageMetaData) (keccakTypes.Challenge, error) {
+func (v *PreimageVerifier) CreateChallenge(ctx context.Context, blockHash common.Hash, oracle VerifierPreimageOracle, preimage keccakTypes.LargePreimageMetaData) (keccakTypes.Challenge, error) {
+	root, err := oracle.GetProposalTreeRoot(ctx, batching.BlockByHash(blockHash), preimage.LargePreimageIdent)
+	if err != nil {
+		return keccakTypes.Challenge{}, fmt.Errorf("failed to get proposal merkle root: %w", err)
+	}
+	if valid, ok := v.knownValid.Get(root); ok && valid {
+		// We've already determined that the keccak transition is valid.
+		// Note that the merkle tree may have been validated by a different proposal but since the tree root
+		// commits to all the input data and the resulting keccak state matrix, any other proposal with the same
+		// root must also have the same inputs and correctly applied keccak.
+		// It is possible that this proposal can't be squeezed because the claimed data length doesn't match the
+		// actual length but the contracts enforce that and it can't be challenged on that basis.
+		return keccakTypes.Challenge{}, matrix.ErrValid
+	}
```

Comment:
Thought - would it be possible to map the `LargePreimageMetaData` here to if the root is valid? Could save us a few rpc calls since we wouldn't need to query for the proposal tree root. The `LargePreimageMetaData` contains a UUID that must be unique and it's hashed so it should be a valid identifier for the root. Is that right?

Reply:
The key thing about the merkle root is that if there's a reorg and data changes, the merkle root will change.  We don't want to cache validity based on the game identity but on its actual content - otherwise we have to start dealing with invalidating caches on reorgs etc.

Reply:
Ah that's right, those pesky reorgs